servers/configuration.go

```diff
@@ -0,0 +1,86 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package servers
+
+import (
+	"crypto/tls"
+
+	"github.com/greymatter-io/nautls/builders"
+	"github.com/pkg/errors"
+)
+
+// Configuration provides a serializable representation of a tls.Config structure for servers.
+type Configuration struct {
+
+	// Authorities defines the trusted certificate authorities for verifying mTLS clients. The values must be URLs that
+	// point to the location of PEM encoded certificates.
+	//
+	// Note that in addition to those schemes supported by [getter](https://godoc.org/github.com/hashicorp/go-getter) a
+	// "base64" scheme is supported for providing the PEM encoded certifiate in the path of the URL directly. This is most
+	// applicable when the certificate data must be provided via an environement variable.
+	Authorities []string `json:"authorities" mapstructure:"authorities" yaml:"authorities"`
+
+	// Certificate defines the server certificate. The value must be a URL that points to the location of a PEM encoded
+	// certificate.
+	//
+	// Note that in addition to those schemes supported by [getter](https://godoc.org/github.com/hashicorp/go-getter) a
+	// "base64" scheme is supported for providing the PEM encoded certifiate in the path of the URL directly. This is most
+	// applicable when the certificate data must be provided via an environement variable.
+	Certificate string `json:"certificate" mapstructure:"certificate" yaml:"certificate"`
+
+	// Key defines the server key. The value must be a URL that points to the location of a PEM encoded key.
+	//
+	// Note that in addition to those schemes supported by [getter](https://godoc.org/github.com/hashicorp/go-getter) a
+	// "base64" scheme is supported for providing the PEM encoded certifiate in the path of the URL directly. This is most
+	// applicable when the certificate data must be provided via an environement variable.
+	Key string `json:"key" mapstructure:"key" yaml:"key"`
+
+	// Authentication defines the client authentication mode for mTLS connections.
+	//
+	// For serialization puposes (i.e., JSON and YAML) the value must be the string representation of a tls.ClientAuthType
+	// constant (e.g., "RequireAnyClientCert"). See https://golang.org/pkg/crypto/tls/#ClientAuthType.
+	Authentication Authentication `json:"authentication" mapstructure:"authentication" yaml:"authentication"`
+}
+
+// TLS returns a tls.Config instance from the configuration. Note that invoking this method on a nil instance is not an
+// error and returns nil.
+func (c *Configuration) TLS() (*tls.Config, error) {
+
+	if c == nil {
+		return nil, nil
+	}
+
+	certificates, err := builders.BuildCertificates(c.Certificate, c.Key)
+	if err != nil {
+		return nil, errors.Wrap(err, "error building certificates")
+	}
+
+	if len(certificates) == 0 {
+		return nil, nil
```

Comment:
Why return nil here? If no certificates are configured, but an authorities pool is, this will error return a nil *tls.Config, which doesn't seem like expected behavior.

Reply:
So the certificate pool is used to verify the client certs when used by a server. Therefore, a server that doesn't have its own certificates (and therefore cannot do TLS) would be asking a client for TLS certs.  Not a valid use case for a server in my mind.  

That said, this does not sit well with me just the same.  I think in practice I should probably return nil only if the Configuration is nil and in all other conditions return a tls.Config as described by the configuration.  Whether the configuration is valid or not will be determined down stream.

Reply:
Should it return nil even if the configuration is nil? Since the client configuration doesn't, might make more sense for consistency's sake if this returned the default tls Config as well. Plus, then a user of this wouldn't have to check if the returned config was nil, as well as the error, on every call to TLS().

Reply:
So the client exposes `HTTP()` and `TLS()` methods.  The `TLS()` method does return `nil`.  The `HTTP()` method does not.  The user has to check anyway because the user needs to know whether to start a tls server or a plaintext server (they are not the same methods).  So we have to provide a way for the configuration to indicate that the server should be plain text or tls.  Nil is the only option and it's very hard to check whether a tls.Config is in a zero state because it has a `Time` field and it causes havoc on equality checking.

Reply:
Gotcha. I also didn't realize that creating an http.Transport with a nil *tls.Client is perfectly valid.